mongo_service.go

```diff
@@ -46,14 +47,41 @@ func (service *MongoService) closeSession() {
 	service.session.Close()
 }
 
-func (service *MongoService) isCurrentNodeMaster() bool {
+func (service *MongoService) isNodeForBackup() bool {
 	db := service.session.DB(defaultDb)
 	result := make(map[string]interface{})
 	err := db.Run(bson.M{"isMaster": 1}, result)
 	if err != nil {
 		log.Panic("Can't check if node is master, db.isMaster() fails", err.Error(), err)
 	}
-	return result["ismaster"].(bool)
+
+	if result["me"] == nil || result["hosts"] == nil {
+		warn.Println("Node is not part of a cluster.")
+		return false
+	}
+
+	master := result["ismaster"].(bool)
+	if master {
+		info.Println("This node is master.")
+		return false
+	}
+
+	thisNode := result["me"].(string)
+	primary := result["primary"].(string)
+	allNodes := result["hosts"].([]string)
+	sort.Strings(allNodes)
+	var lowestSecondary string
+	if allNodes[0] == primary {
+		lowestSecondary = allNodes[1]
+	} else {
+		lowestSecondary = allNodes[0]
+	}
```

Comment:
Currently we have a singleton mongodb-backup service that might run on any
MongoDB node. I'm anticipating changing this to 3 instances of the service,
one running on each MongoDB node. The one that runs on the MongoDB master
has to shut down (as it currently does - so in this event, currently no
backup is performed on this run) and all but one of the instances running
on secondary nodes should also shut down.

It doesn't have to be the same machine on each run (if, for example, AWS
moves machines around and they end up with IPs that order the host names
differently, or there's a MongoDB election that results in a new primary
node); it's just a simple way of making sure service instances running on
different nodes at the same time would order the nodes unambiguously and
come to the same conclusion about which instance should continue to perform
the backup (so, barring exceptional circumstances where no secondary is
available, a backup should always be performed).

On 3 August 2016 at 12:22, peteclark-ft <notifications@github.com> wrote:

> In mongo_service.go
> <https://github.com/Financial-Times/coco-mongodb-backup/pull/9#discussion_r73321299>
> :
>
> > +	master := result["ismaster"].(bool)
> > +	if master {
> > +		info.Println("This node is master.")
> > +		return false
> > +	}
> > +
> > +	thisNode := result["me"].(string)
> > +	primary := result["primary"].(string)
> > +	allNodes := result["hosts"].([]string)
> > +	sort.Strings(allNodes)
> > +	var lowestSecondary string
> > +	if allNodes[0] == primary {
> > +		lowestSecondary = allNodes[1]
> > +	} else {
> > +		lowestSecondary = allNodes[0]
> > +	}
>
> So we're sorting the host:port strings returned in the 'hosts' field, and
> the highest on that list that isn't master is the node to back up on? Is
> this for consistency, so that backups are usually performed on the same
> node?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Financial-Times/coco-mongodb-backup/pull/9/files/92e4e6318bd9d4b129678af4f48611993e2f1a6d#r73321299>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ANIG-ZSl0jJXdW1Ce1sR0vhX26pnJI_Tks5qcHnrgaJpZM4JawVV>
> .
>



-- 

Keith Hatton
Dynamic Publishing

Financial Times
Number One Southwark Bridge
London
SE1 9HL
<https://www.ft.com/>
<https://www.facebook.com/financialtimes>  <https://twitter.com/FT>
<http://www.linkedin.com/company/financial-times>
<https://www.instagram.com/financialtimes/?hl=en>
<http://www.youtube.com/user/FinancialTimesVideos>

-- 

------------------------------

*This email was sent by a company owned by Financial Times Group Limited 
("FT Group <http://aboutus.ft.com/corporate-information/#axzz3rajCSIAt>"), 
registered office at Number One Southwark Bridge, London SE1 9HL.  
Registered in England and Wales with company number 879531. This e-mail may 
contain confidential information. If you are not the intended recipient, 
please notify the sender immediately, delete all copies and do not 
distribute it further.  It could also contain personal views which are not 
necessarily those of the FT Group.  We may monitor outgoing or 
incoming emails as permitted by law.*